specs/Spec.P256.PointOps.fst

```diff
@@ -0,0 +1,279 @@
+  let o = 0xffffffff00000000ffffffffffffffffbce6faada7179e84f3b9cac2fc632551 in
+  assert_norm (o < pow2 256); o
+
+let qelem = x:nat{x < order}
+let qadd (x y:qelem) : qelem = (x + y) % order
+let qmul (x y:qelem) : qelem = (x * y) % order
+let qinv (x:qelem) : qelem = M.pow_mod #order x (order - 2)
+
+let ( +^ ) = qadd
+let ( *^ ) = qmul
+
+
+///  Elliptic curve `y^2 = x^3 + a * x + b`
+
+let aff_point = p:tuple2 nat nat{let (px, py) = p in px < prime /\ py < prime}
+let proj_point = p:tuple3 nat nat nat{let (px, py, pz) = p in px < prime /\ py < prime /\ pz < prime}
+
+// let aff_point = felem & felem           // Affine point
+// let proj_point = felem & felem & felem  // Projective coordinates
+
+let a_coeff : felem = (-3) % prime
+let b_coeff : felem =
+  let b = 0x5ac635d8aa3a93e7b3ebbd55769886bc651d06b0cc53b0f63bce3c3e27d2604b in
+  assert_norm (b < prime); b
+
+
+// Base point
+let g_x : felem =
+  let x = 0x6b17d1f2e12c4247f8bce6e563a440f277037d812deb33a0f4a13945d898c296 in
+  assert_norm (x < prime); x
+let g_y : felem =
+  let y = 0x4fe342e2fe1a7f9b8ee7eb4a7c0f9e162bce33576b315ececbb6406837bf51f5 in
+  assert_norm (y < prime); y
+
+let base_point : proj_point = (g_x, g_y, one)
+
+
+let is_on_curve (p:aff_point) : bool =
+  let (x, y) = p in y *% y = x *% x *% x +% a_coeff *% x +% b_coeff
+
+
+let aff_point_at_inf : aff_point = (zero, zero) // not on the curve!
+let point_at_inf : proj_point = (zero, one, zero)
+
+let is_aff_point_at_inf (p:aff_point) : bool =
+  let (x, y) = p in x = zero && y = zero
+
+let is_point_at_inf (p:proj_point) =
+  let (_, _, z) = p in z = 0
+
+
+let to_proj_point (p:aff_point) : proj_point =
+  let (x, y) = p in (x, y, one)
+
+let to_aff_point (p:proj_point) : aff_point =
+  // if is_proj_point_at_inf p then aff_point_at_inf
+  let (px, py, pz) = p in
+  let zinv = finv pz in
+  let x = px *% zinv in
+  let y = py *% zinv in
+  (x, y)
+
+
+///  Point addition in affine coordinates
+
+let aff_point_double (p:aff_point) : aff_point =
+  let (px, py) = p in
+  if is_aff_point_at_inf p then p
+  else begin
+    if py = 0 then aff_point_at_inf
+    else begin
+      let lambda = (3 *% px *% px +% a_coeff) /% (2 *% py) in
+      let rx = lambda *% lambda -% px -% px in
+      let ry = lambda *% (px -% rx) -% py in
+      (rx, ry) end
+  end
+
+let aff_point_add (p:aff_point) (q:aff_point) : aff_point =
+  let (px, py) = p in let (qx, qy) = q in
+  if is_aff_point_at_inf p then q
+  else begin
+    if is_aff_point_at_inf q then p
+    else begin
+      if p = q then aff_point_double p
+      else begin
+        if qx = px then aff_point_at_inf
+        else begin
+          let lambda = (qy -% py) /% (qx -% px) in
+          let rx = lambda *% lambda -% px -% qx in
+          let ry = lambda *% (px -% rx) -% py in
+          (rx, ry)
+        end
+      end
+    end
+  end
+
+
+///  Point addition and doubling in projective coordinates
+
+// Alg 4 from https://eprint.iacr.org/2015/1060.pdf
```

Comment:
No, we assume that property in `specs/lemmas/Spec.P256.Lemmas.fsti` for now, it would be nice to prove it in the future. In the code, we use the projective formulas